tests/integration/test_rerelate.py

```diff
@@ -0,0 +1,165 @@
+            application_name="alertmanager",
+            channel="edge",
+        ),
+        ops_test.model.deploy(
+            "ch:grafana-k8s",
+            application_name="grafana",
+            channel="edge",
+        ),
+        ops_test.model.deploy(
+            "ch:grafana-agent-k8s",
+            application_name="grafana-agent",
+            channel="edge",
+        ),
+        # TODO: add traefik after fetch-lib (there are breaking changes)
+    )
+
+    await asyncio.gather(
+        ops_test.model.add_relation(app_name, tester_app_name),
+        ops_test.model.add_relation(app_name, "alertmanager"),
+        ops_test.model.add_relation(app_name, "grafana:grafana-source"),
+        ops_test.model.add_relation(app_name, "grafana-agent:send-remote-write"),
+    )
+    await ops_test.model.wait_for_idle(status="active", timeout=600)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_relation(ops_test: OpsTest):
+    await asyncio.gather(
+        ops_test.model.applications[app_name].remove_relation("metrics-endpoint", tester_app_name),
+        ops_test.model.applications[app_name].remove_relation("alertmanager", "alertmanager"),
+        ops_test.model.applications[app_name].remove_relation("grafana-source", "grafana"),
+        ops_test.model.applications[app_name].remove_relation(
+            "receive-remote-write", "grafana-agent"
+        ),
+    )
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=600)
+
+
+@pytest.mark.abort_on_fail
+async def test_rerelate(ops_test: OpsTest):
+    await asyncio.gather(
+        ops_test.model.add_relation(app_name, tester_app_name),
+        ops_test.model.add_relation(app_name, "alertmanager"),
+        ops_test.model.add_relation(app_name, "grafana:grafana-source"),
+        ops_test.model.add_relation(app_name, "grafana-agent:send-remote-write"),
+    )
+    await ops_test.model.wait_for_idle(status="active", timeout=600)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_related_app(ops_test: OpsTest):
+    await asyncio.gather(
+        ops_test.model.applications[tester_app_name].remove(),
+        ops_test.model.applications["alertmanager"].remove(),
+        ops_test.model.applications["grafana"].remove(),
+        ops_test.model.applications["grafana-agent"].remove(),
+    )
+    logger.debug("Applications removed. Blocking for 60 seconds then force removing...")
+    # Block until it is really gone. Added after an itest failed when tried to redeploy:
+    # juju.errors.JujuError: ['cannot add application "...": application already exists']
+    try:
+        await ops_test.model.block_until(
+            lambda: tester_app_name not in ops_test.model.applications,
+            lambda: "alertmanager" not in ops_test.model.applications,
+            lambda: "grafana" not in ops_test.model.applications,
+            lambda: "grafana-agent" not in ops_test.model.applications,
+            timeout=300,
+        )
```

Comment:
`remove` is an alias to `destroy()`, so we may as well use that, though libjuju uses them inconsistently.

Doing this from a cross-model relation may also avoid this (or at least put the tester/agent there), since you can just destroy that model and create a new one.

That said: 

```python
await asyncio.gather(
    ops_test.model.remove_application(tester_app_name, block_until_done=True),
    ops_test.model.remove_application("alertmanager", block_until_done=True),
    ...
)
```

Cleans up the code here a little bit, with the cost of not being able to specify the timeout. 

I imagine that we're actually going to check the data inside Prometheus at some point in these tests, right? Otherwise this is just exercising the negative case from the Juju bug which took us a month. That's ok, but we don't need an explicit test for it here, because we'd see it elsewhere.

But this specific case is actually one of the Juju bugs which only seems to come up in GH runners. Is this actually verifying anything which breaking/re-creating the relation doesn't already do?

Reply:
- Yeah should probably add reldata checks at some point in the future.
- The loki itest is also [waiting](https://github.com/canonical/loki-k8s-operator/blob/d29139883a65149b910cc2f558a3cc4a623ff150/tests/integration/test_rerelate.py#L130) for all apps to be removed, and is also not using `-new` in the app name.
- itest here now fail so I'll add the forced removal like in loki.

Reply:
> Is this actually verifying anything which breaking/re-creating the relation doesn't already do?
The intention here is to capture any possible differences in event sequencing or reldata stucture between remove relation and remove app (without manually remove relation first). It's a bit more "holistic" I guess and as a side effect trips on the juju bug.